### DISCUSSION TITLE
Support for UI created custom models

### DISCUSSION THREAD

**Discussion by dmulyalin** [netbox-community/netbox#12202] [Ideas]
Hello,

There are multiple customization capabilities available in Netbox - on one side of the spectrum we have custom fields that can be used to extend existing models, on the other side we have full blown plugins that can be used to create a lot of things to suite ones use cases. The letter one is very simple to work with, the former one is a coding exercise that depending on the use case might require substantial amount of efforts.

Thought it would be good to have something in between, like capability to add singular custom models to serve the cases where it is desirable to document things in Netbox that are hard to document by using custom fields only, but is an overkill for plugins in terms of efforts required to develop and maintain them. 

For example:

- SR-MPLS artifacts like SID, SRGB, Start Label - can be documented using a single model with reference to device and IP addresses
- ISIS attributes e.g. metrics, afi, area, level, NET ID - can be a model with reference to device, similar for OSPF attributes
- Any other protocols e.g. RSVP, BFD, LDP, PCEP, BGP and their attributes can be documented in similar way
- IPSEC or any other tunnels with their attributes - another singular table/model
- Devices' software - could be a custom field but having a dedicated model would allow to attach other attributes like packages, bug fixes/patches required, known issues etc.
- Assets documentation like device spare parts, cables, sfp modules etc - can be a single model/table that will refer to site or location in one of the fields

All above of course can be covered by using custom plugins, but it would be much easier if there would be a UI driven way to address above and other similar use cases that require only one model to implement. Sort of a cookie cutter way, where UI table would have a list of objects with add/delete/edit buttons  and each object will have identical view to display main attributes, custom fields, tags and that is probably it.

**Comment by endreszabo** [comment]
config contexts?

**Comment by dmulyalin** [comment]
config contexts are not searchable, linkable, apiable, filterable etc. They serve the purpose, but fall short for any granular interaction type of activities.

**Comment by candlerb** [comment]
There's a place I've seen this before, and it's Salesforce ([custom objects](https://help.salesforce.com/s/articleView?id=sf.dev_objectcreate_task_parent.htm&type=5)).

I think it would add a ton of complexity to Netbox, especially since currently each model requires a separate database table, so creating a new custom model would involve creating tables. That's why plugins have their own migrations.

You could instead have all custom models dumped into a single table, storing id plus type plus a JSON column for everything else.  If that's the approach, then someone could write a "custom model plugin" which implements this as proof-of-concept.  It would have a lot of work to do though, as all the form rendering and table rendering would have to be data-driven, showing the correct fields and columns for every different custom model type.

The JSON column would be comparable to `custom_field_data` on a normal model. This suggests another approach: having an essentially empty custom model, with custom fields for all attributes.  This could re-use a lot of the custom field logic. However, *all* custom models would then share the same set of custom fields.

There was a feature request for being able filter the set of custom fields on different instances of the same model. The idea which appeals to me most is to extend the "role" concept to more models, and then the set of custom fields displayed can be filtered by the combination of model + role.  (This would be cool: imagine a device with role "firewall" showing only the custom fields that relate to firewalls).

Salesforce has this too: they call it "[record types](https://help.salesforce.com/s/articleView?id=sf.customize_recordtype.htm&type=5)".
